src/Components/Components/src/Routing/Router.cs

```diff
@@ -155,12 +181,55 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
+        private async Task RunOnNavigateAsync(string path, bool isNavigationIntercepted)
+        {
+            // If this router instance does not provide an OnNavigateAsync parameter
+            // then we render the component associated with the route as per usual.
+            if (OnNavigateAsync == null)
+            {
+                Refresh(isNavigationIntercepted);
+                return;
+            }
+
+            // Create a new CTS for this invocation of the task and
+            // extract the reference to the current one.
+            var pendingOnNavigateCts = _onNavigateCts;
+            var onNavigateCts = new CancellationTokenSource();
+            _onNavigateCts = onNavigateCts;
+
+            var navigateContext = new OnNavigateArgs(path, _onNavigateCts);
+            var task = OnNavigateAsync(navigateContext);
+
+            // If we've already invoked a task and stored its CTS, then
+            // cancel the existing task.
+            if (pendingOnNavigateCts != null)
+            {
+                pendingOnNavigateCts.Cancel();
+            }
+
+            // Create a cancellation task based on the cancellation token
+            // associated with the current running task.
+            var cancellationTaskSource = new TaskCompletionSource();
+            navigateContext.CancellationTokenSource.Token.Register(() =>
+                cancellationTaskSource.TrySetCanceled(navigateContext.CancellationTokenSource.Token));
```

Comment:
Rather than setting up a new TCS, doing the `Register` etc., would it also work to just call `onNavigateCts.Cancel()` on around L223 (just after the `await`)?